lib/srv/termmanager.go

```diff
@@ -122,56 +122,36 @@ func (g *TermManager) Write(p []byte) (int, error) {
 }
 
 func (g *TermManager) Read(p []byte) (int, error) {
-	if len(g.remaining) > 0 {
-		n := copy(p, g.remaining)
-		g.remaining = g.remaining[n:]
-		return n, nil
-	}
-
-	q := make(chan struct{})
-	defer close(q)
-	c := make(chan bool)
-	go func() {
-		g.readStateUpdate.L.Lock()
-		defer g.readStateUpdate.L.Unlock()
-
-		for {
-			g.mu.Lock()
-			on := g.on
-			g.mu.Unlock()
-
-			select {
-			case c <- on:
-			case <-q:
-				close(c)
-				return
-			}
-
-			g.readStateUpdate.Wait()
-		}
-	}()
+	g.mu.Lock()
+	on := g.on
+	g.mu.Unlock()
 
-	on := <-c
 	for {
-		if !on {
+		// Loop until data flow is enabled and there is data to return, or the session is terminated.
+		for !on || len(g.remaining) == 0 {
 			select {
 			case <-g.closed:
 				return 0, io.EOF
-			case on = <-c:
-				continue
+			case on = <-g.readStateUpdate:
```

Comment:
Maybe make readStateUpdate a `chan struct{}`? It doesn't look like we should (nor do) rely on the value we get from it, but instead grab the locks and read g.on instead.